OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,371 @@
+			else
+			{
+				AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), Irc.Client.Nickname, text));
+				Irc.Client.SendMessage(SendType.Message, "#" + Game.Settings.Irc.Channel, text);
+			}
+
+			return true;
+		}
+
+		void OnConnected(object sender, EventArgs e)
+		{
+			AddChatLine("Connected.");
+
+			if (!Rfc2812.IsValidNickname(Game.Settings.Irc.Nickname))
+			{
+				AddChatLine("Invalid nickname. Can't login.");
+				return;
+			}
+
+			Irc.Client.Login(new[] { Game.Settings.Irc.Nickname }, "in-game IRC client", 0, "OpenRA");
+
+			Irc.Client.RfcJoin("#" + Game.Settings.Irc.Channel);
+		}
+
+		void OnError(object sender, ErrorEventArgs e)
+		{
+			AddChatLine("Error: ".F(e.ErrorMessage));
+			Game.RunAfterTick(() =>
+			{
+				Log.Write("irc", e.ToString());
+			});
+		}
+
+		void OnRawMessage(object sender, IrcEventArgs e)
+		{
+			Game.RunAfterTick(() =>
+			{
+				Log.Write("irc", e.Data.RawMessage);
+			});
+		}
+
+		void OnChannelMessage(object sender, IrcEventArgs e)
+		{
+			AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), e.Data.Nick, e.Data.Message));
+		}
+
+		void OnJoin(object sender, JoinEventArgs e)
+		{
+			AddChatLine("{0} joined channel {1}.".F(e.Who, e.Channel));
+			channel = Irc.Client.GetChannel(e.Channel);
+			SyncNicknamePanel();
+		}
+
+		void OnChannelActiveSynced(object sender, IrcEventArgs e)
+		{
+			channel = Irc.Client.GetChannel(e.Data.Channel);
+
+			AddChatLine("{0} users online".F(channel.Users.Count));
+
+			if (!string.IsNullOrEmpty(channel.Topic))
+				AddChatLine("*** Topic: {0}".F(channel.Topic));
+
+			SyncNicknamePanel();
+		}
+
+		void OnNickChange(object sender, NickChangeEventArgs e)
+		{
+			AddChatLine("{0} is now known as {1}.".F(e.OldNickname, e.NewNickname));
+			SyncNicknamePanel();
+		}
+
+		void SyncNicknamePanel()
+		{
+			if (channel == null)
+				return;
+
+			var users = channel.Users;
+			Game.RunAfterTick(() =>
+			{
+				nicknamePanel.RemoveChildren();
+
+				foreach (DictionaryEntry user in users)
+				{
+					var channeluser = (ChannelUser)user.Value;
+					var prefix = channeluser.IsOp ? "@" : channeluser.IsVoice ? "+" : "";
+					var newChild = MakeLabelWidget(nicknameTemplate, prefix + channeluser.Nick);
+					nicknamePanel.AddChild(newChild);
+				}
+			});
+		}
+
+		void OnQuit(object sender, QuitEventArgs e)
+		{
+			AddChatLine("{0} quit.".F(e.Who, e.Data.Channel));
+			channel = Irc.Client.GetChannel(e.Data.Channel);
+			SyncNicknamePanel();
+		}
+
+		void OnPart(object sender, PartEventArgs e)
+		{
+			AddChatLine("{0} left {1}.".F(e.Who, e.Data.Channel));
+			channel = Irc.Client.GetChannel(e.Data.Channel);
+			SyncNicknamePanel();
+		}
+
+		void Disconnect()
+		{
+			if (Irc.Client == null)
+				return;
+
+			Irc.Client.RfcQuit(Game.Settings.Irc.QuitMessage);
+
+			AddChatLine("Disconnecting from {0}...".F(Irc.Client.Address));
+			Irc.Client.Disconnect();
```

Comment:
If you're not connected here (for example, due to an error) this will throw with "The connection could not be disconnected because there is no active connection".

I was able to recreate this by just connecting and disconnecting really fast until the IRC server threw an error at me rather than connecting me.